### PR TITLE
Update PR title once done with the command-issued version update (#131 redo)

### LINKED ISSUE #130: Update PR title after `update version` commands
I've been trying to update the PR title of the "conda forge admin update version" commands but I found a blocker that would require a bigger refactor and stopped there.

Basically I wanted to call this:

```python
def update_pr_title(
    repo_owner: str, repo_name: str, pr_number: int , found_version: str
) -> bool:
    try:
        gh = github.Github(os.environ["GH_TOKEN"])
        repo = gh.get_repo(f"{repo_owner}/{repo_name}")
        pr = repo.get_pull(pr_number)
    except Exception:
        LOGGER.exception(
            "error while trying to get PR title for %s/%s#%s",
            repo_owner,
            repo_name,
            pr_number,
        )
        return False
    if pr.title == "ENH: update package version":  # user didn't change the default
        try:
            pr.edit(title=f"{pr.title} to {found_version}")
        except Exception:
            LOGGER.exception(
                "error while trying to change PR title for %s/%s#%s",
                repo_owner,
                repo_name,
                pr_number,
            )
            return False
    return True
```

right after 

https://github.com/conda-forge/webservices-dispatch-action/blob/7b13ec3da8da6b42ecd76f80b755da1308218b67/webservices_dispatch_action/version_updater.py#L167-L175

But that involves calling GH_TOKEN dependent code, and we can't do that due to the hardening? I thought maybe we can do it over here instead:

https://github.com/conda-forge/webservices-dispatch-action/blob/7b13ec3da8da6b42ecd76f80b755da1308218b67/webservices_dispatch_action/__main__.py#L211-L215

But we aren't returning the found version from the subcommand call (maybe we could return via JSON payloads?), so we wouldn't know what to change the PR title to. 

Am I overthinking this or can we just call `Github()` stuff inside `version_updater.py`?

### PR DESCRIPTION
**Pull request by jaimergp**
- Supersedes https://github.com/conda-forge/webservices-dispatch-action/pull/131
- Closes #130

### CODE REVIEW

File: webservices_dispatch_action/__main__.py [hunk 1]
```diff
@@ -165,6 +165,8 @@ def main():
                     feedstock_dir,
                     "--repo-name",
                     repo_name,
+                    "--pr-number",
+                    pr_num,
```
Comment (beckermr) [inline webservices_dispatch_action/__main__.py]:
```suggestion
                    str(pr_num),
```

**Comment by beckermr** [comment]
See the error on the version update job

```
2024-09-27 17:52:43,428 ERROR    webservices_dispatch_action.version_updater || error while trying to get PR title for conda-forge/cf-autotick-bot-test-package-feedstock#483
Traceback (most recent call last):
  File "/home/runner/work/_actions/conda-forge/webservices-dispatch-action/version-update-pr-title/webservices_dispatch_action/version_updater.py", line 129, in update_pr_title
    gh = github.Github(os.environ["GH_TOKEN"])
                       ~~~~~~~~~~^^^^^^^^^^^^
  File "<frozen os>", line 679, in __getitem__
KeyError: 'GH_TOKEN'
```

https://github.com/conda-forge/cf-autotick-bot-test-package-feedstock/actions/runs/11075355037/job/30776117735

This will go away once we remove the subprocess call.

**Comment by beckermr** [comment]
Also yay for our previous attempt at isolation working for the case we designed it for!

**Comment by jaimergp** [comment]
Hm, any reason we haven't set `concurrency.cancel-in-progress: true` in the GHA workflows??

**Comment by beckermr** [comment]
If we cancel in progress jobs, we can leave the repos that we are testing with in an intermediate state. The tests might be robust enough to turn this off now, but I am not 100% sure.

**Review by beckermr — CHANGES_REQUESTED** [review]
Let's clean out the old entrypoint and main function as well.

File: webservices_dispatch_action/__main__.py [hunk 2]
```diff
@@ -158,30 +159,14 @@ def main():
...
-                if input_version:
-                    cmd += ["--input-version", input_version]
-                LOGGER.info(f"Running command {' '.join(cmd)}")
-                ret = subprocess.run(
-                    cmd,
-                    env=os.environ,
+                    input_version,
+                )
+                version_changed, version_error, found_version = update_version(
+                    feedstock_dir, repo_name, input_version=input_version
```
Comment (beckermr) [inline webservices_dispatch_action/__main__.py]: The first argument needs to be the git_repo object, not an str.

Comment (beckermr) [inline webservices_dispatch_action/__main__.py]:
```suggestion
                    git_repo, repo_name, input_version=input_version
```

**Comment by jaimergp** [comment]
Looks like that worked, but I don't think we are testing the PR title change code paths because of this line:

https://github.com/conda-forge/webservices-dispatch-action/blob/a97415031972fa549fc0535163cd989f42a63291/webservices_dispatch_action/version_updater.py#L144

Any ideas? Or shall we debug that live?
Maybe:

- Add `[TESTING DO NOT MERGE] Version update live test` to the allowed PR titles
- Then add some assertions in `tests/run_live_version_update_test.py`
- And include a PR title revert to the original in the `try/finally` clause?

**Comment by beckermr** [comment]
We should be testing the title change in the tests. So I think we should change the pr title before we do the workflow dispatch, let the version update code change the title, test that the new title is correct, and then put back the original title.

File: webservices_dispatch_action/version_updater.py [hunk 3]
```diff
@@ -132,45 +121,38 @@ def update_version(git_repo, repo_name, input_version=None):
...
-    else:
-        sys.exit(0)
+def update_pr_title(
+    repo_name: str, pr_number: int, found_version: str
+) -> tuple[bool, bool]:
+    """
+    Returns [whether title changed, errored]
+    """
+    try:
+        with sensitive_env():
```
Comment (beckermr) [inline webservices_dispatch_action/version_updater.py:136]: You need to use the functions in the api utils to get the right token.

**Review by beckermr — CHANGES_REQUESTED** [review]

File: webservices_dispatch_action/version_updater.py (hunk 3 above)
Comment (beckermr) [reply webservices_dispatch_action/version_updater.py:136]: See this file: https://github.com/conda-forge/webservices-dispatch-action/blob/main/webservices_dispatch_action/api_sessions.py

Comment (beckermr) [reply webservices_dispatch_action/version_updater.py:136]: This is still not right. The code uses a different token depending on which one is set. You should use the get actor token function and then feed that to function to make the api.

Comment (beckermr) [reply webservices_dispatch_action/version_updater.py:136]: Never mind. This looks good!

**Review by beckermr — APPROVED** [review]

**Comment by jaimergp** [comment]
Looks like it works now! 

image

Thank you for all the guidance! <3

**Comment by beckermr** [comment]
Thank you for the PR!

**Comment by jakirkham** [comment]
Thanks Jaime and Matt! 🙏

This is a great idea 😄

Am wondering if it would also make sense to open PRs as drafts. The we could convert them to ready for review once all changes are pushed

**Comment by beckermr** [comment]
We'll need changes in the webservices repo to but yes we can do that.

**Pull request merged**